### PR TITLE
Enhanced Pipeline RuleUI

### PR DESCRIPTION
**Pull request by gally47**
- [x] [Extractors replication](https://github.com/Graylog2/graylog-plugin-enterprise/issues/5413)
- [x] [Add help icon](https://github.com/Graylog2/graylog-plugin-enterprise/issues/5414) ->  https://github.com/Graylog2/graylog2-server/pull/15883
- [x] [Searchable select](https://github.com/Graylog2/graylog-plugin-enterprise/issues/5414)
- [x] [More detailed simulation](https://github.com/Graylog2/graylog-plugin-enterprise/issues/5415)
- [x] [Referencing → Choose the output of which step to use](https://github.com/Graylog2/graylog-plugin-enterprise/issues/5416)
- [ ] [Reorder](https://github.com/Graylog2/graylog-plugin-enterprise/issues/5417) ？

/nocl

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: graylog2-web-interface/src/components/rules/RuleSimulation.tsx [hunk 1]
```diff
@@ -40,33 +41,40 @@ const MessageShowContainer = styled.div`
...
+
+const OutputContainer = styled.div`
+  margin-top: 8px;
+`;
+
 type Props = {
   rule?: RuleType | RuleBuilderRule,
 };
 
 const RuleSimulation = ({ rule: currentRule }: Props) => {
```
Comment (ousmaneo) [inline graylog2-web-interface/src/components/rules/RuleSimulation.tsx:56]: When I copy a message and use it as the value in the Rule simulation. The simulation result reset with the default simulation message. See below.
rule-simulation

**Review by ousmaneo — CHANGES_REQUESTED** [review]
Looks great! I found some issue, and some nitpick below. 
Submitting this part now, and continue testing.

**Review by ousmaneo — CHANGES_REQUESTED** [review]
I found some console errors, and also some issues from testing some function conditions and actions. Sometimes errors happen from validation as we have seen for `grok` and proper error is not displayed to the user.

A part from that the UI is great !

File: graylog2-web-interface/src/components/rules/rule-builder/RuleBlockForm.tsx [hunk 2]
```diff
@@ -67,6 +70,19 @@ const SelectedBlockInfo = styled(Row)(({ theme }) => css`
...
+const OptionTitle = styled.p(({ theme }) => css`
+  margin-bottom: ${theme.spacings.xxs};
+`);
+
+const OptionDescription = styled.p<{ $isSelected: boolean }>(({ theme, $isSelected }) => css`
+  color: ${$isSelected ? theme.colors.gray[90] : theme.colors.gray[50]};
+  margin-bottom: ${theme.spacings.xxs};
+`);
+
 const RuleBlockForm = ({
```
[the remark this answers is not in the document]
Comment (grotlue) [reply graylog2-web-interface/src/components/rules/rule-builder/RuleBlockForm.tsx:86]: We are aware of this one, but currently, we need both forms for the HTML validation to work. It's on our list to see if we can restructure the UI, but for now, we decided to keep it as it's "just" a warning und not breaking anything.

File: graylog2-web-interface/src/components/rules/rule-builder/RuleBuilderBlock.tsx [hunk 3]
```diff
@@ -53,7 +51,7 @@ const RuleBuilderBlock = ({
   blockDict,
```
[the remark this answers is not in the document]
Comment (grotlue) [reply graylog2-web-interface/src/components/rules/rule-builder/RuleBuilderBlock.tsx:51]: > When creating the condition `field_contains`, the block display does not show a description only the name of the function. See below
> creation:
> image
> Result:
> image

> This also happens for other string-related conditions. Is it normal? Note that in the current version it is showing the field we want to check and the value we are checking.

@moesterheld Are we missing some text/headlines for those?

File: graylog2-web-interface/src/components/rules/rule-builder/RuleBuilder.tsx [hunk 4]
```diff
@@ -26,35 +26,68 @@ import { ConfirmDialog, FormSubmit } from 'components/common';
...
+const SimulatorSwitchContaner = styled.div`
+  display: flex;
+  align-items: center;
+  position: absolute;
+  top: 4px;
+  right: 16px;
+  z-index: 1;
+`;
+
 const RuleBuilder = () => {
```
[the remark this answers is not in the document]
Comment (grotlue) [reply graylog2-web-interface/src/components/rules/rule-builder/RuleBuilder.tsx:89]: That should be gone now.

**Review by ousmaneo — CHANGES_REQUESTED** [review]
I did some more test, and the GROK-related function still throws some errors; when that happens, an empty error is thrown. The Rule block is then empty as you see in the screenshot below.
image

Can we handle this error so we show the user that validation failed?

**Review by ousmaneo — APPROVED** [review]
I tested again after some unsupported functions have been disabled and everything works fine.

LGTM! Thank you!

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/IsNull.java [hunk 5]
```diff
@@ -50,6 +51,10 @@ public FunctionDescriptor<Boolean> descriptor() {
                 .returnType(Boolean.class)
                 .params(of(valueParam))
                 .description("Checks whether a value is 'null'")
+                .ruleBuilderEnabled(false)
+                .ruleBuilderName("Check if not null")
+                .ruleBuilderTitle("Check if value is not null")
```
Comment (patrickmann) [inline graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/IsNull.java:56]: should be `is null` not `is not null`

**Comment by patrickmann** [comment]
These functions lack a primary parameter:
IsUrl, IsPeriod, Debug, IsJson, SetField, Syslog*Conversion, IsUrl

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/conversion/IsMap.java [hunk 6]
```diff
@@ -48,6 +49,10 @@ public FunctionDescriptor<Boolean> descriptor() {
                 .returnType(Boolean.class)
                 .params(valueParam)
                 .description("Checks whether a value is a map")
+                .ruleBuilderEnabled(false)
+                .ruleBuilderName("Check if map")
+                .ruleBuilderTitle("Check if '${value}' is a map")
```
Comment (patrickmann) [inline graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/conversion/IsMap.java:54]: [Grammar](https://www.merriam-webster.com/grammar/if-vs-whether-difference-usage#:~:text=For%20clarity%2C%20it%20is%20best,clauses%2C%20phrases%2C%20or%20words.) OCD comment: _For clarity, it is best to use whether in reference to a choice or alternatives ("we're going whether it rains or not") and if when establishing a condition ("we will go if it doesn't rain")._

Using _if_ implies the meaning: If value is map perform some unspecified check. _Whether_ expresses the intended meaning: Determine whether value is map or not map.

I'm fine keeping _if_ here - it's not really ambiguous and it's shorter.

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/DropMessage.java [hunk 7]
```diff
@@ -53,6 +54,10 @@ public FunctionDescriptor<Void> descriptor() {
                         messageParam
                 ))
                 .description("Discards a message from further processing, after completing the current stage")
+                .ruleBuilderEnabled()
+                .ruleBuilderName("Drop message")
+                .ruleBuilderTitle("Remove message from the processing pipeline after rule")
```
Comment (patrickmann) [inline graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/DropMessage.java:59]: Prefer "after current stage", since there is often confusion as to when exactly the message gets dropped.

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/RemoveFromStream.java [hunk 8]
```diff
@@ -99,7 +100,11 @@ public FunctionDescriptor<Void> descriptor() {
                         nameParam,
                         idParam,
                         messageParam))
-                .description("Removes a message from a stream. Removing the last stream will put the message back onto the default stream. To complete drop a message use the drop_message function.")
+                .description("Removes a message from a stream. Removing the last stream will put the message back onto the default stream. To complete drop a message use the drop_message function. If no specific message is provided, it removes the message from the currently processed message.")
```
Comment (patrickmann) [inline graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/RemoveFromStream.java:103]: "To complete**ly** drop"

I don't understand the last sentence:  "If no specific message is provided, it removes the message from the currently processed message."

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/strings/Capitalize.java [hunk 9]
```diff
@@ -43,4 +44,16 @@ protected boolean isLocaleAware() {
     protected String apply(String value, Locale unused) {
         return StringUtils.capitalize(value);
     }
+
+    @Nonnull
+    @Override
+    protected String getRuleBuilderName() {
+        return "Capitalize String";
```
Comment (patrickmann) [inline graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/strings/Capitalize.java:51]:
```suggestion
        return "Capitalize string";
```

**Comment by patrickmann** [comment]
`Convert to source code` does not notice when I click `Update rule`. The converted rule remains the previously saved version - I have to do `Update rule & close` to see the last changes.

File: graylog2-server/src/test/java/org/graylog/plugins/pipelineprocessor/rulebuilder/parser/ConditionParserTest.java [hunk 10]
```diff
@@ -151,12 +150,12 @@ public void generate_WhenRuleConditionsContainsMultipleFunctions() {
...
-                  && function1(
+        assertThat(conditionParser.generate(steps, RuleBuilderStep.Operator.AND, 1)).isEqualTo("""
+                  function1(
                     required : "val1"
                   )
-                  && ! function2()
+                    AND
+                  ! function2()
                 """.stripTrailing());
     }
```
Comment (patrickmann) [inline graylog2-server/src/test/java/org/graylog/plugins/pipelineprocessor/rulebuilder/parser/ConditionParserTest.java:161]: Add a TODO comment to `generate_WhenRuleConditionsContainsOneFragment` or delete it.

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/conversion/IsMap.java (hunk 6 above)
Comment (moesterheld) [reply graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/conversion/IsMap.java:54]: Keeping it for the name, changing it for the title

**Comment by moesterheld** [comment]
> These functions lack a primary parameter: IsUrl, IsPeriod, Debug, IsJson, SetField, Syslog*Conversion, IsUrl

Added the primary parameter for all functions which already have rule builder metadata (e.g. not for the syslog ones)

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/rulebuilder/db/migrations/V20230720161500_AddExtractorFragments.java [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    public ZonedDateTime createdAt() {
+        return ZonedDateTime.parse("2023-06-13T15:44:00Z");
+    }
+
+    @Override
+    public void upgrade() {
+        log.debug("Adding extractor fragments via migration");
+        if (Objects.nonNull(clusterConfigService.get(MigrationCompleted.class))) {
+            log.debug("Migration already completed!");
+//            return;
```
[the remark this answers is not in the document]
Comment (moesterheld) [reply graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/rulebuilder/db/migrations/V20230720161500_AddExtractorFragments.java:60]: After discussion with Maxwell, we decided to at least leave these commented out until next cycle for the 5.2 release because likely there might still be changes. I'll add a TO-DO to one of them

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/rulebuilder/db/migrations/V20230720161500_AddExtractorFragments.java [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+                        .ruleBuilderEnabled()
+                        .ruleBuilderName("Extract lookup value")
+                        .ruleBuilderTitle("Extract value for field '${field}', do lookup in '${lookupTable}' and set value to new field '${newField}'")
+                        .ruleBuilderFunctionGroup(RuleBuilderFunctionGroup.EXTRACTORS)
+                        .build())
+                .fragmentOutputVariable(resultvariable)
+                .build();
+    }
+
+    private void addFragment(RuleFragment ruleFragment) {
```
[the remark this answers is not in the document]
Comment (moesterheld) [reply graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/rulebuilder/db/migrations/V20230720161500_AddExtractorFragments.java:236]: good idea, I'll refactor that

**Review by patrickmann — APPROVED** [review]
Looks good

File: graylog2-web-interface/src/components/rules/rule-builder/ConfirmNavigateToSourceCodeEditorModal.tsx [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+      <Modal.Header closeButton>
+        <Modal.Title>You have unsaved changes</Modal.Title>
+      </Modal.Header>
+      <Modal.Body>
+        Do you want to save changes before navigating to Source Code Editor?
+      </Modal.Body>
+      <Modal.Footer>
+        <Button type="button"
+                bsStyle="success"
+                onClick={async () => {
```
Comment (ousmaneo) [inline graylog2-web-interface/src/components/rules/rule-builder/ConfirmNavigateToSourceCodeEditorModal.tsx:54]: Can we extract this to a function?

File: graylog2-web-interface/src/components/rules/rule-builder/ConfirmNavigateToSourceCodeEditorModal.tsx [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+                    ...rule,
+                    title: rule.title || autoSavedTitle,
+                  });
+
+                  history.push(Routes.SYSTEM.PIPELINES.RULE('new'));
+                }}>
+          Save
+        </Button>
+        <Button type="button"
+                onClick={() => {
```
Comment (ousmaneo) [inline graylog2-web-interface/src/components/rules/rule-builder/ConfirmNavigateToSourceCodeEditorModal.tsx:73]: Can we also extract this to a function?

**Review by ousmaneo — APPROVED** [review]
Tested and LGTM!
I have added two small suggestions.
